index.js

```diff
@@ -152,35 +194,41 @@ FastForEach.prototype.processQueue = function () {
 
 
 // Process a changeItem with {status: 'added', ...}
-FastForEach.prototype.added = function (index, value) {
+FastForEach.prototype.added = function (changeItem) {
+  var index = changeItem.index;
+  var valuesToAdd = changeItem.isBatch ? changeItem.value : [changeItem.value];
   var referenceElement = this.lastNodesList[index - 1] || null;
-  var templateClone = this.templateNode.cloneNode(true);
-  var childNodes = ko.virtualElements.childNodes(templateClone);
-  var childContext;
- 
-  if(this.noContext) {
-    childContext = this.$context.extend({
-      '$item' : value
-    });
-  } else {
-    childContext = this.$context.createChildContext(value, this.as || null);
-  }
+  // gather all childnodes for a possible batch insertion
+  var allChildNodes = [];
+
+  for (var i = 0, len = valuesToAdd.length; i < len; ++i) {
+    var templateClone = this.templateNode.cloneNode(true);
+    var childContext;
 
-  this.lastNodesList.splice(index, 0, childNodes[childNodes.length - 1]);
-  ko.applyBindingsToDescendants(childContext, templateClone);
+    if (this.noContext) {
+      childContext = this.$context.extend({
+        '$item': valuesToAdd[i]
+      });
+    } else {
+      childContext = this.$context.createChildContext(valuesToAdd[i], this.as || null);
+    }
 
-  // Nodes are inserted in reverse order - pushed down immediately after
-  // the last node for the previous item or as the first node of element.
-  for (var i = childNodes.length - 1; i >= 0; --i) {
-    var child = childNodes[i];
-    if (!child) return;
-    ko.virtualElements.insertAfter(this.element, child, referenceElement);
+    // apply bindings first, and then process child nodes, because bindings can add childnodes
+    ko.applyBindingsToDescendants(childContext, templateClone);
+
+    var childNodes = ko.virtualElements.childNodes(templateClone);
+    allChildNodes.push.apply(allChildNodes, childNodes);
+    this.lastNodesList.splice(index + i, 0, childNodes[childNodes.length - 1]);
   }
+
+  insertAllAfter(this.element, allChildNodes, referenceElement);
 };
 
 
 // Process a changeItem with {status: 'deleted', ...}
-FastForEach.prototype.deleted = function (index, value) {
+FastForEach.prototype.deleted = function (changeItem) {
+  var index = changeItem.index;
+  var value = changeItem.value;
```

Comment:
Because you call these methods dynamically at queue processing, so chaning the "added" method implicated changing the "deleted" too, to make both remain same signature.